esp32_s3_usb_otg/esp32_s3_usb_otg.c

```diff
@@ -0,0 +1,430 @@
+    disp_drv.flush_cb = lvgl_port_flush_callback;
+    disp_drv.draw_buf = &disp_buf;
+    disp_drv.user_data = panel_handle;
+    return lv_disp_drv_register(&disp_drv);
+}
+
+static void lvgl_port_tick_increment(void *arg)
+{
+    /* Tell LVGL how many milliseconds has elapsed */
+    lv_tick_inc(LVGL_TICK_PERIOD_MS);
+}
+
+static esp_err_t lvgl_port_tick_init(void)
+{
+    // Tick interface for LVGL (using esp_timer to generate periodic event)
+    const esp_timer_create_args_t lvgl_tick_timer_args = {
+        .callback = &lvgl_port_tick_increment,
+        .name = "LVGL tick"
+    };
+    esp_timer_handle_t lvgl_tick_timer = NULL;
+    ESP_ERROR_CHECK(esp_timer_create(&lvgl_tick_timer_args, &lvgl_tick_timer));
+    return esp_timer_start_periodic(lvgl_tick_timer, LVGL_TICK_PERIOD_MS * 1000);
+}
+
+static void lvgl_port_task(void *arg)
+{
+    ESP_LOGI(TAG, "Starting LVGL task");
+    while (1) {
+        bsp_display_lock(0);
+        uint32_t task_delay_ms = lv_timer_handler();
+        bsp_display_unlock();
+        if (task_delay_ms > 500) {
+            task_delay_ms = 500;
+        }
+        vTaskDelay(pdMS_TO_TICKS(task_delay_ms));
+    }
+}
+
+lv_disp_t *bsp_display_start(void)
+{
+    lv_init();
+    bsp_display_brightness_init();
+    lv_disp_t *disp = lvgl_port_display_init();
+    lvgl_port_tick_init();
+    lvgl_mux = xSemaphoreCreateMutex();
+    assert(lvgl_mux);
+    xTaskCreate(lvgl_port_task, "LVGL task", 4096, NULL, CONFIG_BSP_DISPLAY_LVGL_TASK_PRIORITY, NULL);
+    return disp;
+}
+
+bool bsp_display_lock(uint32_t timeout_ms)
+{
+    assert(lvgl_mux && "bsp_display_start must be called first");
+
+    const TickType_t timeout_ticks = (timeout_ms == 0) ? portMAX_DELAY : pdMS_TO_TICKS(timeout_ms);
+    return xSemaphoreTake(lvgl_mux, timeout_ticks) == pdTRUE;
+}
+
+void bsp_display_unlock(void)
+{
+    assert(lvgl_mux && "bsp_display_start must be called first");
+    xSemaphoreGive(lvgl_mux);
+}
+
+void bsp_usb_mode_select_device(void)
+{
+    // Make sure the pin is configured
+    const gpio_config_t led_io_config = {
+        .pin_bit_mask = BIT64(BSP_USB_MODE_SEL),
+        .mode = GPIO_MODE_OUTPUT,
+        .pull_up_en = GPIO_PULLUP_DISABLE,
+        .pull_down_en = GPIO_PULLDOWN_DISABLE,
+        .intr_type = GPIO_INTR_DISABLE
+    };
+    ESP_ERROR_CHECK(gpio_config(&led_io_config));
+    ESP_ERROR_CHECK(gpio_set_level(BSP_USB_MODE_SEL, 0));
+}
+
+
+void bsp_usb_mode_select_host(void)
+{
+    // Make sure the pin is configured
+    const gpio_config_t led_io_config = {
+        .pin_bit_mask = BIT64(BSP_USB_MODE_SEL),
+        .mode = GPIO_MODE_OUTPUT,
+        .pull_up_en = GPIO_PULLUP_DISABLE,
+        .pull_down_en = GPIO_PULLDOWN_DISABLE,
+        .intr_type = GPIO_INTR_DISABLE
+    };
+    ESP_ERROR_CHECK(gpio_config(&led_io_config));
+    ESP_ERROR_CHECK(gpio_set_level(BSP_USB_MODE_SEL, 1));
+}
+
+void bsp_usb_host_power_mode(bsp_usb_host_power_mode_t mode, bool limit_500mA)
```

Comment:
I don't know, what about `_init `functions, but this function can be called in runtime and I think, it should return `esp_err_t `and not assert on errors. I think, that all runtime functions should returns errors instead of assert. The code can working next, if this fails.

Reply:
This is a good point for discussion; the BSPs are not targeting production quality code, but rather quick prototyping.

So the idea here is that we don't want to bother users with error-code checking. The app will fail at runtime with clear error message.

What do you think?

Reply:
I understand. It's a pity, that BSP is meant only for quick prototyping and not for production code. I thought, that it can help with developping (they can use part of code). 

When I am develop new device, I am using same parts of schematic as on developer boards. Then I can use the same SW there (e.g. display). It is only my opinion.

But OK, we can leave it.